### PR TITLE
8317562: [JFR] Compilation queue statistics

### PR DESCRIPTION
**Pull request by macarte**
Adding a new periodic jfr event to monitor and output statistics for the compiler queues.  You will see one event per compiler queue (c1 and c2)

Passes tier1 on linux (x86) and mac (aarch64)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317562](https://bugs.openjdk.org/browse/JDK-8317562): [JFR] Compilation queue statistics (**Enhancement** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [6c0b1670](https://git.openjdk.org/jdk/pull/16211/files/6c0b16702c7ba73fd8f88d24e413f6afc7b6f6d4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16211/head:pull/16211` \
`$ git checkout pull/16211`

Update a local copy of the PR: \
`$ git checkout pull/16211` \
`$ git pull https://git.openjdk.org/jdk.git pull/16211/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16211`

View PR using the GUI difftool: \
`$ git pr show -t 16211`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16211.diff">https://git.openjdk.org/jdk/pull/16211.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16211#issuecomment-1768736923)

### CODE REVIEW

File: src/hotspot/share/compiler/compileBroker.hpp [hunk 1]
```diff
@@ -112,6 +118,9 @@ class CompileQueue : public CHeapObj<mtCompiler> {
   bool         is_empty() const                  { return _first == nullptr; }
   int          size()     const                  { return _size;          }
 
+  int         get_peak_size()     const          { return _peak_size; }
```
Comment (karianna) [inline src/hotspot/share/compiler/compileBroker.hpp:121]: whitespace looks out by one?

File: src/hotspot/share/jfr/metadata/metadata.xml [hunk 2]
```diff
@@ -848,6 +848,19 @@
...
+    <Field type="CompilerType" name="compiler" label="Compiler" />
+    <Field type="long" contentType="hertz" name="ingress" label="Ingress" description="Number of added requests per second"/>
+    <Field type="long" contentType="hertz" name="egress" label="Egress" description="Number of removed requests per second"/>
+    <Field type="long" contentType="count" name="size" label="Size" description="Current queue size"/>
+    <Field type="long" contentType="count" name="peak" label="Peak Size" description="Peak queue size"/>
+    <Field type="long" contentType="count" name="added" label="Added" description="Added requests"/>
+    <Field type="long" contentType="count" name="removed" label="Removed" description="Removed requests"/>
+    <Field type="long" contentType="count" name="totalAdded" label="Total Added" description="Total requests"/>
+    <Field type="long" contentType="count" name="totalRemoved" label="Total Removed" description="Total removed requests"/>
+    <Field type="int" name="compilerThreadCount" label="Number of Compiler Threads for this queue" />
```
Comment (karianna) [inline src/hotspot/share/jfr/metadata/metadata.xml:861]:
```suggestion
    <Field type="int" name="compilerThreadCount" label="Number of Compiler Threads for this queue"/>
```

File: src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+#include "jfr/periodic/jfrCompilerQueueUtilization.hpp"
+
+struct CompilerQueueEntry {
+  CompileQueue* compilerQueue;
+  u8 compiler;
+  uint64_t added;
+  uint64_t removed;
+};
+
+// If current counters are less than previous we assume the interface has been reset
```
Comment (karianna) [inline src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp:37]:
```suggestion
// If current counters are less than previous, we assume the interface has been reset
```

File: src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  assert(interval.value() > 0, "invariant");
+  if (current <= old) {
+    return 0;
+  }
+  return ((current - old) * NANOSECS_PER_SEC) / interval.nanoseconds();
+}
+
+void JfrCompilerQueueUtilization::send_events() {
+  static CompilerQueueEntry compilerQueueEntries[2] = {
+    {CompileBroker::c1_compile_queue(), 1, 0,0},
```
Comment (karianna) [inline src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp:49]:
```suggestion
    {CompileBroker::c1_compile_queue(), 1, 0, 0},
```

File: src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    {CompileBroker::c2_compile_queue(), 2, 0,0}};
```
Comment (karianna) [inline src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp:50]:
```suggestion
    {CompileBroker::c2_compile_queue(), 2, 0, 0}};
```

File: src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+  const JfrTicks cur_time = JfrTicks::now();
+  static JfrTicks last_sample_instant;
+  const JfrTickspan interval = cur_time - last_sample_instant;
+  for(int i = 0; i < 2; i ++)
```
Comment (karianna) [inline src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp:55]: 2 is a magic number, maybe a comment above the for declartion explaining why?

**Review by karianna — CHANGES_REQUESTED** [review]
very nitpicky - take or leave the suggestions :-)

File: src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp (hunk 6 above)
Comment (macarte) [reply src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp]: Thanks for the feedback, made this clearer

File: test/jdk/jdk/jfr/event/compiler/TestCompilerQueueUtilization.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+import jdk.jfr.Recording;
+import jdk.jfr.consumer.RecordedEvent;
+import jdk.test.lib.jfr.EventNames;
+import jdk.test.lib.jfr.Events;
+
+/**
+ * @test
+ * @key jfr
+ * @requires vm.hasJFR
```
Comment (JohnTortugo) [inline test/jdk/jdk/jfr/event/compiler/TestCompilerQueueUtilization.java:36]: Does it need C1 and/or C2?

File: src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+// If current counters are less than previous, we assume the interface has been reset
+// If no bytes have been either sent or received, we'll also skip the event
+static uint64_t rate_per_second(uint64_t current, uint64_t old, const JfrTickspan& interval) {
+  assert(interval.value() > 0, "invariant");
+  if (current <= old) {
+    return 0;
+  }
+  return ((current - old) * NANOSECS_PER_SEC) / interval.nanoseconds();
+}
+
```
Comment (JohnTortugo) [inline src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp:55]: NIT: Too many new lines?

File: src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+void JfrCompilerQueueUtilization::send_events() {
+  static CompilerQueueEntry compilerQueueEntries[num_compiler_queues] = {
+    {CompileBroker::c1_compile_queue(), c1_compiler_queue_id, &CompileBroker::get_c1_thread_count, 0, 0},
+    {CompileBroker::c2_compile_queue(), c2_compiler_queue_id, &CompileBroker::get_c2_thread_count, 0, 0}};
+
+  const JfrTicks cur_time = JfrTicks::now();
+  static JfrTicks last_sample_instant;
+  const JfrTickspan interval = cur_time - last_sample_instant;
+  for(int i = 0; i < num_compiler_queues; i ++)
+  {
```
Comment (JohnTortugo) [inline src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp:67]: I think this doesn't follow the coding style. See document in "/doc/" folder in the repository.

File: src/hotspot/share/compiler/compileBroker.hpp [hunk 10]
```diff
@@ -88,6 +88,9 @@ class CompileQueue : public CHeapObj<mtCompiler> {
   CompileTask* _first_stale;
 
   int _size;
+  int _total_added;
```
Comment (JohnTortugo) [inline src/hotspot/share/compiler/compileBroker.hpp:91]: Can total be negative?

**Review by JohnTortugo — COMMENTED** [review]
Some quick comments.

File: src/hotspot/share/compiler/compileBroker.hpp (hunk 10 above)
Comment (macarte) [reply src/hotspot/share/compiler/compileBroker.hpp]: no - changed to uint

File: test/jdk/jdk/jfr/event/compiler/TestCompilerQueueUtilization.java (hunk 7 above)
Comment (macarte) [reply test/jdk/jdk/jfr/event/compiler/TestCompilerQueueUtilization.java:36]: yes requires c1 or c2

Comment (macarte) [reply test/jdk/jdk/jfr/event/compiler/TestCompilerQueueUtilization.java:36]: added requirement that java is NOT run with Xint (interpreter mode)

File: src/hotspot/share/compiler/compileBroker.hpp [hunk 11]
```diff
@@ -112,6 +118,9 @@ class CompileQueue : public CHeapObj<mtCompiler> {
   bool         is_empty() const                  { return _first == nullptr; }
   int          size()     const                  { return _size;          }
 
+  int         get_peak_size()     const          { return _peak_size; }
+  int         get_total_added()   const          { return _total_added; }
+  int         get_total_removed() const          { return _total_removed; }
```
Comment (vnkozlov) [inline src/hotspot/share/compiler/compileBroker.hpp:123]: Fields are `uint` type. These accessors should also return `uint`.

File: src/hotspot/share/compiler/compileBroker.cpp [hunk 12]
```diff
@@ -515,6 +520,14 @@ CompileQueue* CompileBroker::compile_queue(int comp_level) {
...
 }
 
+CompileQueue* CompileBroker::c1_compile_queue() {
+  return _c1_compile_queue;
+}
+
+CompileQueue* CompileBroker::c2_compile_queue() {
+  return _c2_compile_queue;
+}
+
```
Comment (vnkozlov) [inline src/hotspot/share/compiler/compileBroker.cpp:530]: Note, `*_compiler_queue` could be `nullptr` if VM is build without C2 or C1 or when run with `-XX:-TieredCompilation` (only C2 is used) or with `-XX:TierdStopAtLevel={1,2,3}` (only C1 is used).

Make sure you check it in JFR event.

**Review by vnkozlov — COMMENTED** [review]
Would be interesting to test it with Graal which use `c2_compile_queue` (Graal JIT replaces C2).

File: src/hotspot/share/compiler/compileBroker.cpp (hunk 12 above)
Comment (macarte) [reply src/hotspot/share/compiler/compileBroker.cpp:530]: Thank you!  The JFR event does check for NULL (now nullptr)

**Review by vnkozlov — APPROVED** [review]
Good for compiler part of changes.

File: test/jdk/jdk/jfr/event/compiler/TestCompilerQueueUtilization.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+ * @run main/othervm jdk.jfr.event.compiler.TestCompilerQueueUtilization
+ */
+public class TestCompilerQueueUtilization {
+    private final static String EVENT_NAME = EventNames.CompilerQueueUtilization;
+
+    public static void main(String[] args) throws Exception {
+        Recording recording = new Recording();
+        recording.enable(EVENT_NAME);
+        recording.start();
+        recording.stop();
```
Comment (brianjstafford) [inline test/jdk/jdk/jfr/event/compiler/TestCompilerQueueUtilization.java:48]: Would it be useful to delay the recording stop in order to give more of an opportunity for compiler events to occur?

File: test/jdk/jdk/jfr/event/compiler/TestCompilerQueueUtilization.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2013, 2023, Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (macarte) [reply test/jdk/jdk/jfr/event/compiler/TestCompilerQueueUtilization.java]: copy paste mistake (fixed)

File: src/hotspot/share/jfr/metadata/metadata.xml [hunk 15]
```diff
@@ -848,6 +848,19 @@
     <Field type="float" contentType="hertz" name="switchRate" label="Switch Rate" description="Number of context switches per second" />
   </Event>
 
+  <Event name="CompilerQueueUtilization" category="Java Virtual Machine, Compiler" label="Compiler Queue Utilization" period="everyChunk">
+    <Field type="CompilerType" name="compiler" label="Compiler" />
+    <Field type="long" contentType="hertz" name="ingress" label="Ingress" description="Number of added requests per second"/>
+    <Field type="long" contentType="hertz" name="egress" label="Egress" description="Number of removed requests per second"/>
+    <Field type="long" contentType="count" name="size" label="Size" description="Current queue size"/>
+    <Field type="long" contentType="count" name="peak" label="Peak Size" description="Peak queue size"/>
+    <Field type="long" contentType="count" name="added" label="Added" description="Added requests"/>
```
[the remark this answers is not in the document]
Comment (macarte) [reply src/hotspot/share/jfr/metadata/metadata.xml]: -removed contenttType="count"
-removed superfluous descriptions

File: test/jdk/jdk/jfr/event/compiler/TestCompilerQueueUtilization.java (hunk 13 above)
Comment (macarte) [reply test/jdk/jdk/jfr/event/compiler/TestCompilerQueueUtilization.java]: I've followed the pattern for other events that output periodically, the test will fail if there are no events due to the line 'Events.hasEvents(events)

File: test/jdk/jdk/jfr/event/compiler/TestCompilerQueueUtilization.java [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+ * @requires vm.hasJFR
+ * @requires vm.compMode!="Xint"
+ * @library /test/lib
+ * @run main/othervm jdk.jfr.event.compiler.TestCompilerQueueUtilization
+ */
+public class TestCompilerQueueUtilization {
+    private final static String EVENT_NAME = EventNames.CompilerQueueUtilization;
+
+    public static void main(String[] args) throws Exception {
+        Recording recording = new Recording();
```
[the remark this answers is not in the document]
Comment (macarte) [reply test/jdk/jdk/jfr/event/compiler/TestCompilerQueueUtilization.java]: Added

File: src/hotspot/share/jfr/metadata/metadata.xml [hunk 17]
```diff
@@ -848,6 +848,19 @@
...
+    <Field type="CompilerType" name="compiler" label="Compiler" />
+    <Field type="long" contentType="hertz" name="ingress" label="Ingress" description="Number of added requests per second"/>
+    <Field type="long" contentType="hertz" name="egress" label="Egress" description="Number of removed requests per second"/>
+    <Field type="long" name="size" label="Size"/>
+    <Field type="long" name="peak" label="Peak Size"/>
+    <Field type="long" name="added" label="Added"/>
+    <Field type="long" name="removed" label="Removed"/>
+    <Field type="long" name="totalAdded" label="Total Added"/>
+    <Field type="long" name="totalRemoved" label="Total Removed"/>
+    <Field type="int" name="compilerThreadCount" label="Number of Compiler Threads for this queue"/>
```
Comment (egahlin) [inline src/hotspot/share/jfr/metadata/metadata.xml:861]: The label should be short and use headline-style capitalization, how about "Compiler Thread Count"?
https://docs.oracle.com/en/java/javase/21/docs/api/jdk.jfr/jdk/jfr/Label.html

File: src/hotspot/share/jfr/metadata/metadata.xml [hunk 18]
```diff
@@ -848,6 +848,19 @@
     <Field type="float" contentType="hertz" name="switchRate" label="Switch Rate" description="Number of context switches per second" />
   </Event>
 
+  <Event name="CompilerQueueUtilization" category="Java Virtual Machine, Compiler" label="Compiler Queue Utilization" period="everyChunk">
+    <Field type="CompilerType" name="compiler" label="Compiler" />
+    <Field type="long" contentType="hertz" name="ingress" label="Ingress" description="Number of added requests per second"/>
```
Comment (egahlin) [inline src/hotspot/share/jfr/metadata/metadata.xml:853]: Could we use other terms then ingress and egress? Something that is more in general use,

File: src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+};
+
+// If current counters are less than previous, we assume the interface has been reset
+// If no bytes have been either sent or received, we'll also skip the event
+static uint64_t rate_per_second(uint64_t current, uint64_t old, const JfrTickspan& interval) {
+  assert(interval.value() > 0, "invariant");
+  if (current <= old) {
+    return 0;
+  }
+  return ((current - old) * NANOSECS_PER_SEC) / interval.nanoseconds();
```
Comment (egahlin) [inline src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp:53]: Shouldn't it be ticks per second here?

File: src/hotspot/share/jfr/metadata/metadata.xml [hunk 20]
```diff
@@ -848,6 +848,19 @@
...
+    <Field type="long" contentType="hertz" name="egress" label="Egress" description="Number of removed requests per second"/>
+    <Field type="long" name="size" label="Size"/>
```
Comment (egahlin) [inline src/hotspot/share/jfr/metadata/metadata.xml:855]: The label could be more clarifying? Queue Size?

File: src/hotspot/share/jfr/metadata/metadata.xml [hunk 21]
```diff
@@ -848,6 +848,19 @@
...
+    <Field type="long" name="peak" label="Peak Size"/>
```
Comment (egahlin) [inline src/hotspot/share/jfr/metadata/metadata.xml:856]: peakQueueSize?

File: src/hotspot/share/jfr/metadata/metadata.xml (hunk 18 above)
Comment (macarte) [reply src/hotspot/share/jfr/metadata/metadata.xml]: Thought about enqueue/dequeue, but went with added/removed so its consistent with the other fields

File: src/hotspot/share/jfr/metadata/metadata.xml (hunk 17 above)
Comment (macarte) [reply src/hotspot/share/jfr/metadata/metadata.xml]: Fixed, thank you for the reference

File: src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp (hunk 19 above)
Comment (macarte) [reply src/hotspot/share/jfr/periodic/jfrCompilerQueueUtilization.cpp:53]: This resolves to ticks per second; could replace with (current - old) / interval.seconds() but this introduces floats.  This was taken from NetworkUtilization, I assume they went this way to keep the math in integers

**Comment by MBaesken** [comment]
Hi , we are seeing now the following error in our tests (with this PR added)  👍 
java.lang.RuntimeException: Field ingress not in struct
	at jdk.test.lib.Asserts.fail(Asserts.java:634)
	at jdk.test.lib.jfr.Events.getValueDescriptor(Events.java:154)
	at jdk.test.lib.jfr.Events.assertField(Events.java:69)
	at jdk.jfr.event.compiler.TestCompilerQueueUtilization.main(TestCompilerQueueUtilization.java:55)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
	at java.base/java.lang.Thread.run(Thread.java:1570)
java.lang.RuntimeException: Field ingress not in event
	at jdk.test.lib.Asserts.fail(Asserts.java:634)
	at jdk.test.lib.jfr.Events.assertField(Events.java:81)
	at jdk.jfr.event.compiler.TestCompilerQueueUtilization.main(TestCompilerQueueUtilization.java:55)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
	at java.base/java.lang.Thread.run(Thread.java:1570)

**Comment by macarte** [comment]
> Hi , we are seeing now the following error in our tests (with this PR added) when running jdk/jfr/event/compiler/TestCompilerQueueUtilization.java
> 
> java.lang.RuntimeException: Field ingress not in struct at jdk.test.lib.Asserts.fail(Asserts.java:634) at jdk.test.lib.jfr.Events.getValueDescriptor(Events.java:154) at jdk.test.lib.jfr.Events.assertField(Events.java:69) at jdk.jfr.event.compiler.TestCompilerQueueUtilization.main(TestCompilerQueueUtilization.java:55) at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) at java.base/java.lang.reflect.Method.invoke(Method.java:580) at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138) at java.base/java.lang.Thread.run(Thread.java:1570) java.lang.RuntimeException: Field ingress not in event at jdk.test.lib.Asserts.fail(Asserts.java:634) at jdk.test.lib.jfr.Events.assertField(Events.java:81) at jdk.jfr.event.compiler.TestCompilerQueueUtilization.main(TestCompilerQueueUtilization.java:55) at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) at java.base/java.lang.reflect.Method.invoke(Method.java:580) at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138) at java.base/java.lang.Thread.run(Thread.java:1570)

Thank you for bringing this to my attention, I had failed to commit the updated test (to reflect the field name changes), it's not part of tier1 tests and so the github checks passed

**Comment by MBaesken** [comment]
With the updated test file,  the jtreg test error is gone.

**Comment by macarte** [comment]
@egahlin / @vnkozlov / @MBaesken / @brianjstafford / @JohnTortugo / @karianna - thank you for your reviews and feedback!

**Pull request closed without merging**